DOC: add a tutorial of scipy.optimize.linprog (#11847)
scipy/scipy: doc/source/tutorial/optimize.rst

   algorithms (e.g., hybrid Powell, Levenberg-Marquardt or large-scale
   methods such as Newton-Krylov [KK]_).

> 6. Linear programming (:func:`linprog`), minimize a linear objective function

Author: I think based on "Basic" and "interpolate" tutorial, "optimize" should use ".. contents::" for link to each section. So, I did it.

sides of the second inequality by :math:`-1` to get:

.. math::
        x_1 -x_2 -3x_3 + 0x_4  &\leq 5\\
        -2x_1 + 3x_2 + 7x_3 - 3x_4 &\leq -10\\

These equations can be converted to matrix form:

.. math::
    A_{ub} x \leq b_{ub}\\

where

.. math::
   :nowrap:

    \begin{equation*} A_{ub} =
    \begin{bmatrix} 1 & -1 & -3 & 0 \\
                    -2 & 3 & 7 & -3
    \end{bmatrix}
    \end{equation*}

.. math::
   :nowrap:

    \begin{equation*} b_{ub} =
    \begin{bmatrix} 5 \\
                    -10
    \end{bmatrix}
    \end{equation*}

Next, let's consider the two equality constraints. Showing zero weights explicitly, these are:

.. math::
        2x_1 + 8x_2 + 1x_3 + 0x_4 &= 60\\
        4x_1 + 4x_2 + 0x_3 + 1x_4 &= 60\\

These equations can be converted to matrix form:

.. math::
    A_{eq} x = b_{eq}\\

where

.. math::
   :nowrap:

    \begin{equation*} A_{eq} =
    \begin{bmatrix} 2 & 8 & 1 & 0 \\
                    4 & 4 & 0 & 1
    \end{bmatrix}
    \end{equation*}

.. math::
   :nowrap:

    \begin{equation*} b_{eq} =
    \begin{bmatrix} 60 \\
                    60
    \end{bmatrix}
    \end{equation*}

> Lastly, let's consider the separate inequality constraints on each decision variable, which are known as

Reviewer: There were some objective things left (at the bottom), I figured I'd add some more subjective improvements.

"box constraints" or "simple bounds". These constraints can be applied using the bounds argument of :func:`linprog`.
As noted in the :func:`linprog` documentation, the default value of bounds is ``(0, None)``, meaning that the
lower bound on each decision variable is 0, and the upper bound on each decision variable is infinity:
all the decision variables are non-negative. Our bounds are different, so we will need to specify the lower and upper bound on each
decision variable as a tuple and group these tuples into a list.

Finally, we can solve the transformed problem using :func:`linprog`.

::

    >>> import numpy as np
    >>> from scipy.optimize import linprog
    >>> c = np.array([-29.0, -45.0, 0.0, 0.0])
    >>> A_ub = np.array([[1.0, -1.0, -3.0, 0.0],

>     ...                [-2.0, 3.0, 7.0, -3.0]])

Reviewer: ```suggestion
    ...                 [-2.0, 3.0, 7.0, -3.0]])
```

    >>> b_ub = np.array([5.0, -10.0])
    >>> A_eq = np.array([[2.0, 8.0, 1.0, 0.0],
    ...                 [4.0, 4.0, 0.0, 1.0]])
    >>> b_eq = np.array([60.0, 60.0])
    >>> x0_bounds = (0, None)
    >>> x1_bounds = (0, 5.0)
    >>> x2_bounds = (-np.inf, 0.5)  # +/- np.inf can be used instead of None
    >>> x3_bounds = (-3.0, None)
    >>> bounds = [x0_bounds, x1_bounds, x2_bounds, x3_bounds]
    >>> result = linprog(c, A_ub=A_ub, b_ub=b_ub, A_eq=A_eq, b_eq=b_eq, bounds=bounds)
    >>> print(result)
         con: array([15.5361242 , 16.61288005])  # may vary
         fun: -370.2321976308326  # may vary
     message: 'The algorithm terminated successfully and determined that the problem is infeasible.'

>          nit: 6

Reviewer: ```suggestion
         nit: 6  # may vary
```

       slack: array([ 0.79314989, -1.76308532])  # may vary
      status: 2
     success: False
           x: array([ 6.60059391,  3.97366609, -0.52664076,  1.09007993])  # may vary

The result states that our problem is infeasible, meaning that there is no solution vector that satisfies all the
constraints. That doesn't necessarily mean we did anything wrong; some problems truly are infeasible.
Suppose, however, that we were to decide that our bound constraint on :math:`x_1` was too tight and that it could be loosened
to :math:`0 \leq x_1 \leq 6`. After adjusting our code ``x1_bounds = (0, 6)`` to reflect the change and executing it again:

::

    >>> x1_bounds = (0, 6)
    >>> bounds = [x0_bounds, x1_bounds, x2_bounds, x3_bounds]
    >>> result = linprog(c, A_ub=A_ub, b_ub=b_ub, A_eq=A_eq, b_eq=b_eq, bounds=bounds)
    >>> print(result)
        con: array([9.78840831e-09, 1.04662945e-08])  # may vary
        fun: -505.97435889013434  # may vary
    message: 'Optimization terminated successfully.'

>         nit: 4

Reviewer: ```suggestion
        nit: 4  # may vary
```

The result shows the optimization was successful.
We can check the objective value (``result.fun``) is same as :math:`c^Tx`:

::

    >>> x = np.array(result.x)
    >>> print(c @ x)
    -505.97435889013434  # may vary

We can also check that all constraints are satisfied within reasonable tolerances:

::

    >>> print(b_ub - (A_ub @ x).flatten())  # this is equivalent to result.slack
    [ 6.52747190e-10, -2.26730279e-09]  # may vary
    >>> print(b_eq - (A_eq @ x).flatten())  # this is equivalent to result.con
    [ 9.78840831e-09, 1.04662945e-08]]  # may vary
    >>> print([0 <= result.x[0], 0 <= result.x[1] <= 6.0, result.x[2] <= 0.5, -3.0 <= result.x[3]])
    [True, True, True, True]

If we need greater accuracy, typically at the expense of speed, we can solve using the ``revised simplex`` method:

::

    >>> result = linprog(c, A_ub=A_ub, b_ub=b_ub, A_eq=A_eq, b_eq=b_eq, bounds=bounds, method='revised simplex')
    >>> print(result)
        con: array([0.00000000e+00, 7.10542736e-15])  # may vary
        fun: -505.97435897435895    # may vary
    message: 'Optimization terminated successfully.'

>         nit: 5

Reviewer: ```suggestion
        nit: 5  # may vary
```

    \begin{bmatrix} 2 & 8 & 1 & 0 \\
                    4 & 4 & 0 & 1
    \end{bmatrix}
    \end{equation*}

.. math::
   :nowrap:

    \begin{equation*} b_{eq} =
    \begin{bmatrix} 60 \\
                    60
    \end{bmatrix}
    \end{equation*}

Lastly, let's consider the separate inequality constraints on each decision variable, which are known as
"box constraints" or "simple bounds". These constraints can be applied using the bounds argument of :func:`linprog`.
As noted in the :func:`linprog` documentation, the default value of bounds is ``(0, None)``, meaning that the
lower bound on each decision variable is 0, and the upper bound on each decision variable is infinity:
all the decision variables are non-negative. Our bounds are different, so we will need to specify the lower and upper bound on each
decision variable as a tuple and group these tuples into a list.

Finally, we can solve the transformed problem using :func:`linprog`.

::

    >>> import numpy as np
    >>> from scipy.optimize import linprog
    >>> c = np.array([-29.0, -45.0, 0.0, 0.0])
    >>> A_ub = np.array([[1.0, -1.0, -3.0, 0.0],
    ...                [-2.0, 3.0, 7.0, -3.0]])

>     >>> b_ub = np.array([5.0, -10.0]),

Reviewer: ```suggestion
    >>> b_ub = np.array([5.0, -10.0])
```

Reviewer: After fixing this, fix the outputs below.

Author: I'm sorry, which output do you mean? result one?

Author: I fixed it. Thank you gain.

    ...                [-2.0, 3.0, 7.0, -3.0]])
    >>> b_ub = np.array([5.0, -10.0]),
    >>> A_eq = np.array([[2.0, 8.0, 1.0, 0.0],
    ...                 [4.0, 4.0, 0.0, 1.0]])
    >>> b_eq = np.array([60.0, 60.0])
    >>> x0_bounds = (0, None)
    >>> x1_bounds = (0, 5.0)
    >>> x2_bounds = (-np.inf, 0.5)  # +/- np.inf can be used instead of None
    >>> x3_bounds = (-3.0, None)
    >>> bounds = [x0_bounds, x1_bounds, x2_bounds, x3_bounds]
    >>> result = linprog(c, A_ub=A_ub, b_ub=b_ub, A_eq=A_eq, b_eq=b_eq, bounds=bounds)
    >>> print(result)
         con: array([15.5361242 , 16.61288005])  # may vary
         fun: -370.2321976308326  # may vary
     message: 'The algorithm terminated successfully and determined that the problem is infeasible.'
         nit: 6
       slack: array([ 0.79314989, -1.76308532])  # may vary
      status: 2
     success: False
           x: array([ 6.60059391,  3.97366609, -0.52664076,  1.09007993])  # may vary

The result states that our problem is infeasible, meaning that there is no solution vector that satisfies all the
constraints. That doesn't necessarily mean we did anything wrong; some problems truly are infeasible.

> Suppose, however, that we decided that our bound constraint on :math:x_1 were too tight and that it could be loosened

Reviewer: ```suggestion
Suppose, however, that we were to decide that our bound constraint on :math:x_1 was too tight and that it could be loosened
```

> to :math:0 \leq x_1 \leq 6\\. After adjusting our code x1_bounds = (0, 6) to reflect the change and executing it again:

Reviewer: Math/code rendering

Author: I fixed it

::

    >>> x1_bounds = (0, 6)
    >>> bounds = [x0_bounds, x1_bounds, x2_bounds, x3_bounds]
    >>> result = linprog(c, A_ub=A_ub, b_ub=b_ub, A_eq=A_eq, b_eq=b_eq, bounds=bounds)
    >>> print(result)
        con: array([9.78840831e-09, 1.04662945e-08])  # may vary
        fun: -505.97435889013434  # may vary
    message: 'Optimization terminated successfully.'
        nit: 4
      slack: array([ 6.52747190e-10, -2.26730279e-09])  # may vary
     status: 0
    success: True
          x: array([ 9.41025641,  5.17948718, -0.25641026,  1.64102564])  # may vary

The result shows the optimization was successful.
We can check the objective value (``result.fun``) is same as :math:`c^Tx`:

::

    >>> x = np.array(result.x)
    >>> print(c @ x)
    -505.97435889013434  # may vary

We can also check that all constraints are satisfied within reasonable tolerances:

::

    >>> print(b_ub - A_ub @ x)  # this is equivalent to result.slack

>     [[  5.]  # may vary

Reviewer: results need to match code

Author: I fixed it

in making a simple choice that worked reasonably well, but there is a
lot more depth to this topic than is shown here.

Linear programming (:func:`linprog`)
------------------------------------

The function :func:`linprog` can minimize a linear objective function
subject to linear equality and inequality constraints. This kind of

> problem is well known as Linear programming. Linear programming solves

Author: I fixed.

problems of the following form:

.. math::

        \min_x \ & c^T x \\
        \mbox{such that} \ & A_{ub} x \leq b_{ub},\\
        & A_{eq} x = b_{eq},\\
        & l \leq x \leq u ,

where :math:`x` is a vector of decision variables; :math:`c`, :math:`b_{ub}`,
:math:`b_{eq}`, :math:`l`, and :math:`u` are vectors; and :math:`A_{ub}` and
:math:`A_{eq}` are matrices.

> In this tutorial, we will try to solve some typical linear programming

Author: I fixed.

problems using :func:`linprog`.

Linear programming example
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Consider the following simple linear programming problem:

.. math::
        \max_{x_1, x_2, x_3, x_4} \ & 29x_1 + 45x_2 \\
        \mbox{such that} \
        & x_1 -x_2 -3x_3 \leq 5\\
        & 2x_1 -3x_2 -7x_3 + 3x_4 \geq 10\\
        & 2x_1 + 8x_2 + x_3 = 60\\
        & 4x_1 + 4x_2 + x_4 = 60\\
        & x_i \geq 0 ,

> We need some mathematical deformations to convert the target problem to the form of :func:`linprog`.

Author: I fixed.

First of all, let's consider the objective function. The problem has 4 decision variables :math:`x_1, x_2, x_3, x_4`.

> We define a vector of decision variables :math:`x = [x_1, x_2, x_3, x_4]`. The problem want to maximize the objective

Author: I fixed.

> function, but :func:`linprog` can accept a minimization problem. This is easily remedied by converting the maximize

Author: I fixed.

:math:`29x_1 + 45x_2` to minimizing :math:`-29x_1 -45x_2`. Also, :math:`x_3, x_4` are not shown in the objective

> function. That means the weights elements of the objective function for :math:`x_3, x_4` are zeros. So, the objective

Author: I fixed.

function can be converted to:

.. math::
        \min_{x_1, x_2, x_3, x_4} \ -29x_1 -45x_2 + 0x_3 + 0x_4

By recalling vector multiplication, the objective weights vector :math:`c` of :func:`linprog` in this problem
should be

.. math::
        c = [-29, -45, 0, 0]

Next, let's consider the two inequality constraints. These can be converted by using the "greater than" inequality

> constraint to a "less than" inequality constraint by multiplying both sides by a factor of :math:`-1`:

Author: I fixed.

.. math::
        x_1 -x_2 -3x_3 + 0x_4  &\leq 5\\
        -2x_1 + 3x_2 + 7x_3 - 3x_4 &\leq -10\\

> These equations can be converted as matrix form:

Author: I fixed.

.. math::
    A_{ub} x \leq b_{ub}\\

where

.. math::
   :nowrap:

    \begin{equation*} A_{ub} =
    \begin{bmatrix} 1 & -1 & -3 & 0 \\
                    -2 & 3 & 7 & -3
    \end{bmatrix}
    \end{equation*}

.. math::
   :nowrap:

    \begin{equation*} b_{ub} =
    \begin{bmatrix} 5 \\
                    -10
    \end{bmatrix}
    \end{equation*}

> Next, let's consider the two equality constraints. These can be converted like:

Author: I fixed.

.. math::
        2x_1 + 8x_2 + 1x_3 + 0x_4 &= 60\\
        4x_1 + 4x_2 + 0x_3 + 1x_4 &= 60\\

> These equations can be converted as matrix form:

Reviewer: ```suggestion
These equations can be converted to matrix form:
```

Author: I fixed.

.. math::
    A_{eq} x = b_{eq}\\

where

.. math::
   :nowrap:

    \begin{equation*} A_{eq} =
    \begin{bmatrix} 2 & 8 & 1 & 0 \\
                    4 & 4 & 0 & 1
    \end{bmatrix}
    \end{equation*}

.. math::
   :nowrap:

    \begin{equation*} b_{eq} =
    \begin{bmatrix} 60 \\
                    60
    \end{bmatrix}
    \end{equation*}

The last inequality constraint :math:`x_i \geq 0` is minimum value constraint for all decision variables, which can be
applied as the ``bounds`` argument of :func:`linprog`. As you can see the API doc :func:`linprog`, the default values of
``bound`` argument is ``(0, None)`` that means all decision variables are non-negative. So, we don't need to do anything

> the last inequality constraint in this case.

Reviewer: It is good for the tutorial to emphasize that the default is non-negativity, but please add other bounds to demonstrate how they are used.

Author: I added new box constraints.

> Finally, we can solve the desired problem using the mathematical converts and :func:`linprog`.

Reviewer: ```suggestion
Finally, we can solve the transformed problem using :func:`linprog`.
```

Author: I fixed.

::

    from scipy.optimize import linprog
    import numpy as np

    c = np.array([-29.0, -45.0, 0.0, 0.0])
    A_ub = np.array([[1.0, -1.0, -3.0, 0.0],
                    [-2.0, 3.0, 7.0, -3.0]])
    b_ub = np.array([[5.0],
                    [-10.0]])
    A_eq = np.array([[2.0, 8.0, 1.0, 0.0],
                     [4.0, 4.0, 0.0, 1.0]])
    b_eq = np.array([[60.0],
                    [60.0]])

    result = linprog(c, A_ub=A_ub, b_ub=b_ub, A_eq=A_eq, b_eq=b_eq)

The result is

::

    >>> print(result)
    con: array([3.07489501e-09, 3.19440119e-09])
    fun: -500.79999997382595
    message: 'Optimization terminated successfully.'
    nit: 5
    slack: array([ 1.0000000e+00, -1.0003518e-09])
    status: 0
    success: True
    x: array([9.2000000e+00, 5.2000000e+00, 4.3425461e-11, 2.4000000e+00])

::

>     >>> [print("x_", i, ":", np.round(result.x[i], 2)) for i in range(len(result.x))]

Reviewer: I would suggest using `format` rather than actually rounding.

Author: I think this printing function is not needed any more because of the optimization problem change. I delete it.

These equations can be converted to matrix form:

.. math::
    A_{eq} x = b_{eq}\\

where

.. math::
   :nowrap:

    \begin{equation*} A_{eq} =
    \begin{bmatrix} 2 & 8 & 1 & 0 \\
                    4 & 4 & 0 & 1
    \end{bmatrix}
    \end{equation*}

.. math::
   :nowrap:

    \begin{equation*} b_{eq} =
    \begin{bmatrix} 60 \\
                    60
    \end{bmatrix}
    \end{equation*}

Lastly, let's consider the separate inequality constraints on each decision variable, which are known as
"box constraints" or "simple bounds". These constraints can be applied using the bounds argument of :func:`linprog`.
As noted in the :func:`linprog` documentation, the default value of bounds is (0, None), meaning that the
lower bound on each decision variable is 0, and the upper bound on each decision variable is infinity:
all the decision variables are non-negative. Our bounds are different, so we specify the lower and upper bound on each
decision variable as a tuple and group these tuples into a list.

Finally, we can solve the transformed problem using :func:`linprog`.

::

    >>> import numpy as np
    >>> from scipy.optimize import linprog

>     >>> c = np.array([-29.0, -45.0, 0.0, 0.0]).T

Reviewer: ```suggestion
    >>> c = np.array([-29.0, -45.0, 0.0, 0.0])
```

    >>> A_ub = np.array([[1.0, -1.0, -3.0, 0.0],
    ...                [-2.0, 3.0, 7.0, -3.0]])

>     >>> b_ub = np.array([[5.0],

Reviewer: ```suggestion
    >>> b_ub = np.array([5.0, -10.0]),
```
I don't think we should be teaching the reader to make these 2D unnecessarily. Besides, you wrote above that `b_ub` is a vector. In that case, it should be a 1D array, not a an N x 1 (2D) array. Similar for `b_eq` below.

    ...                [-10.0]])
    >>> A_eq = np.array([[2.0, 8.0, 1.0, 0.0],
    ...                 [4.0, 4.0, 0.0, 1.0]])
    >>> b_eq = np.array([[60.0],
    ...                [60.0]])
    >>> x0_bounds = (0, None)
    >>> x1_bounds = (0, 5.0)
    >>> x2_bounds = (-np.inf, 0.5)  # +/- np.inf can be used instead of None
    >>> x3_bounds = (-3.0, 0)
    >>> bounds = [x0_bounds, x1_bounds, x2_bounds, x3_bounds]
    >>> result = linprog(c, A_ub=A_ub, b_ub=b_ub, A_eq=A_eq, b_eq=b_eq, bounds=bounds)
    >>> print(result)
        con: array([42.58855175, 45.54023356])  # may vary
        fun: -143.26958745795812  # may vary
    message: 'The algorithm terminated successfully and determined that the problem is infeasible.'
        nit: 4
      slack: array([  2.89860769, -12.11677835])  # may vary
     status: 2
    success: False
          x: array([ 2.41503507,  1.62741268, -0.43792331, -1.71002454])  # may vary

The result shows our problem is infeasible. It means that there is no solution vector that satisfies all the

> constraints. What did we do wrong? If we check the code carefully, we can notice that we've set the :math:`x_3` bound

Reviewer: ```suggestion
constraints. What did we do wrong?
```
I don't think the part about the bound on x3 is the important mistake. Leave the bound on x3 alone and just correct the bound on x1 to (0, 6.0). Better yet, let's not make the reader look for mistakes. Actually change the bound on x1 above to:
```
0 \leq x_1 \leq 5\\
```

Then I'd suggest something like this. 

The result states that our problem is infeasible, meaning that there is no solution vector that satisfies all the
constraints. That doesn't necessarily mean we did anything wrong; some problems truly are infeasible. Suppose, however, that we decided that our bound constraint on :math:`x_1` were too tight and that it could be loosened to :math:`0 \leq x_1 \leq 6\\`. After adjusting our code `x1_bounds = (0, 6)` to reflect the change and executing it again:

Author: I fixed it.

constraint wrong. We've set the maximum bound for :math:`x_3` to :math:`0`, but it is not needed.

We can solve it with correct bound settings:

::

    >>> x3_bounds = (-3.0, None)
    >>> bounds = [x0_bounds, x1_bounds, x2_bounds, x3_bounds]
    >>> result = linprog(c, A_ub=A_ub, b_ub=b_ub, A_eq=A_eq, b_eq=b_eq, bounds=bounds)
    >>> print(result)
         con: array([9.78840831e-09, 1.04662945e-08])  # may vary
         fun: -505.97435889013434  # may vary
     message: 'Optimization terminated successfully.'  # may vary
         nit: 4
       slack: array([ 6.52747190e-10, -2.26730279e-09])  # may vary
      status: 0
     success: True
           x: array([ 9.41025641,  5.17948718, -0.25641026,  1.64102564])  # may vary

> The result shows the optimization was succeeded.

Reviewer: ```suggestion
The result shows the optimization was successful.
```

We can check the objective value (``result.fun``) is same as :math:`c^Tx`:

::

>     >>> x = np.array([result.x]).T

Reviewer: ```suggestion
    >>> x = np.array(result.x)
```
See next comment.

>     >>> print(c.T @ x)

Reviewer: ```suggestion
    >>> print(c @ x)
```
I would argue that we shouldn't show transposes. Even if it makes it look more like the math, users should know - or be taught - that it's not necessary.

Author: OK. I got it. Let's remove transpose.

    -505.97435889013434  # may vary

> We can also check all constraints are almost all satisfied:

Reviewer: ```suggestion
We can also check that all constraints are satisfied within reasonable tolerances:
```

::

>     >>> print(A_ub @ x)

Reviewer: ```suggestion
    >>> print(b_ub - A_ub @ x)  # this is equivalent to result.slack
```

    [[  5.]  # may vary
    [-10.]]  # may vary
    >>> print(b_ub)
    [[  5.]  # may vary
    [-10.]]  # may vary

>     >>> print(A_eq @ x)

Reviewer: ```suggestion
    >>> print(b_eq - A_eq @ x)  # this is equivalent to result.con
```

    [[59.99999999]  # may vary
    [59.99999999]]  # may vary
    >>> print(b_eq)
    [[60.]  # may vary
    [60.]]  # may vary
    >>> print([0 <= result.x[0], 0 <= result.x[1] <= 6.0, result.x[2] <= 0.5, -3.0 <= result.x[3]])
    [True, True, True, True]

> 

Reviewer: If we need greater accuracy, typically at the expense of speed, we can solve using the `revised simplex` method:

```
>>> result = linprog(c, A_ub=A_ub, b_ub=b_ub, A_eq=A_eq, b_eq=b_eq, 
                     bounds=bounds, method='revised simplex')
>>> print(result)
     con: array([0.00000000e+00, 7.10542736e-15])  # may vary
     fun: -505.97435897435895    # may vary
 message: 'Optimization terminated successfully.'
     nit: 5
   slack: array([ 1.77635684e-15, -3.55271368e-15])  # may vary
  status: 0
 success: True
       x: array([ 9.41025641,  5.17948718, -0.25641026,  1.64102564])  # may vary

Author: I fixed.

        c = [-29, -45, 0, 0]

Next, let's consider the two inequality constraints. These can be converted by using the "greater than" inequality
constraint to a "less than" inequality constraint by multiplying both sides by a factor of :math:`-1`. The first
one is the "less than" inequality constraint, however second one is the "greater than" inequality constraint. So, only

> second one should be converted to align inequality sign direction:

Reviewer: Next, let's consider the two inequality constraints. The first one is a "less than" inequality; however, the second one is a "greater than" inequality. `linprog` only accepts "less than" inequality constraints, so we multiply both sides of the second inequality by :math:`-1` to get:

Author: I fixed.

    \end{bmatrix}
    \end{equation*}

Lastly, let's consider the minimum and maximum inequality constraints for each decision variable, which is known as
"box constraint". These constraints can be applied as the ``bounds`` argument of :func:`linprog`.
As you can see the API doc :func:`linprog`, the default values of ``bounds`` argument is ``(0, None)`` that means
all decision variables are non-negative. Using None indicates that there is no bound. So, we need to set each bound

> as ``bounds`` argument in this case.

Reviewer: Lastly, let's consider the separate inequality constraints on each decision variable, which are known as
"box constraints" or "simple bounds". These constraints can be applied using the ``bounds`` argument of :func:`linprog`.
As noted in the :func:`linprog` documentation, the default value of ``bounds`` is ``(0, None)``, meaning that the 
lower bound on each decision variable is ``0``, and the upper bound on each decision variable is infinity:
all the decision variables are non-negative.  Our bounds are different, so we specify the lower and upper bound on each decision variable as a tuple and group these tuples into a list:
```
x0_bounds = (0, None)
x1_bounds = (0, 5.0)
x2_bounds = (None, 0.5)
x3_bounds = (-3.0, 0)
bounds = [x0_bounds, x1_bounds, x2_bounds, x3_bounds]
```

Author: I fixed.

Finally, we can solve the transformed problem using :func:`linprog`.

::

    >>> from scipy.optimize import linprog
    >>> import numpy as np
    >>> c = np.array([-29.0, -45.0, 0.0, 0.0])
    >>> A_ub = np.array([[1.0, -1.0, -3.0, 0.0],
    ...                [-2.0, 3.0, 7.0, -3.0]])
    >>> b_ub = np.array([[5.0],
    ...                [-10.0]])
    >>> A_eq = np.array([[2.0, 8.0, 1.0, 0.0],
    ...                 [4.0, 4.0, 0.0, 1.0]])
    >>> b_eq = np.array([[60.0],
    ...                [60.0]])
    >>> x0_bounds = (0, None)
    >>> x1_bounds = (0, 5.0)

>     >>> x2_bounds = (None, 0.5)

Reviewer: ```suggestion
    >>> x2_bounds = (-np.inf, 0.5)  # +/- np.inf can be used instead of None
```

Author: I fixed.

    >>> x3_bounds = (-3.0, 0)
    >>> bounds = [x0_bounds, x1_bounds, x2_bounds, x3_bounds]
    >>> result = linprog(c, A_ub=A_ub, b_ub=b_ub, A_eq=A_eq, b_eq=b_eq, bounds=bounds)

The result is

::

    >>> print(result)

>         con: array([42.58855175, 45.54023356])

Reviewer: ```suggestion
        con: array([42.58855175, 45.54023356])  # may vary
```
This is why Travis is failing. It performs a docstring test, so you need to indicate that the numbers can change.

Author: I fixed.

>         fun: -143.26958745795812

Reviewer: ```suggestion
        fun: -143.26958745795812  # may vary
```

Author: I fixed.

    message: 'The algorithm terminated successfully and determined that the problem is infeasible.'
        nit: 4

>       slack: array([  2.89860769, -12.11677835])

Reviewer: ```suggestion
      slack: array([  2.89860769, -12.11677835])  # may vary
```

Author: I fixed.

     status: 2
    success: False

>           x: array([ 2.41503507,  1.62741268, -0.43792331, -1.71002454])

Reviewer: ```suggestion
          x: array([ 2.41503507,  1.62741268, -0.43792331, -1.71002454])  # may vary
```

Author: I fixed.
